conflation/map_matching.py

```diff
@@ -0,0 +1,292 @@
+    for e in resp["edges"][1:-1]:
+        way_length = e["length"]  # Kilometers
+        density_value = e["density"]
+        admin = resp["admins"][e["end_node"]["admin_index"]]
+        country, region = admin["country_code"], admin["state_code"]
+        road_class = e["road_class"]
+        t = e["end_node"]["elapsed_time"]
+        t_elapsed_on_way = t - prev_t  # Seconds
+
+        # The elapsed time should be monotonically increasing. If not, this is a bad match and we will skip it
+        if t < prev_t:
+            print("Skipping b/c time not monotonically increasing {} -> {}".format(prev_t, t))
+            return
+        # If the elapsed time doesn't increase for some reason, we can't make any measurement here, so we will ignore it
+        if t == prev_t:
+            # json.dumps([{'lon': b['lon'], 'lat': b['lat'], 'type': b['type'], 'time': i} for i, b in
+            #             enumerate(reversed(body['shape']))])
+            continue
+
+        kph = way_length / t_elapsed_on_way * 3600
+
+        # Skip measurements that are going too fast
+        if kph > MAXIMUM_SPEED:
+            print("Skipping b/c kph of {} > limit of {}".format(kph, MAXIMUM_SPEED))
+            return
+
+        # Ordered tuple that holds all the information that we need to classify this edge, as well as the speed
+        # calculated. See aggregation.MAP_MATCH_COLS for the meaning of each column
+        edge_data = (classify_density(density_value), road_class, get_type_for_edge(e), kph)
+        add_data_to_map_matches(map_matches, country, region, edge_data)
+
+        prev_t = t
+
+
+def write_map_matches(
+    map_matches_dir: str, map_matches: dict[str, dict[str, list[tuple]]]
+) -> None:
+    """
+    Writes the results from map_matches to disk. Follows a format where each config is located in a dir corresponding to
+    the the iso3166-1 spec, and has a filename corresponding to the iso3166-2 spec.
+    """
+
+    for country, regions in map_matches.items():
+        country_dir = os.path.join(map_matches_dir, country)
+        # Make the dir if it does not exist yet
+        if not os.path.exists(country_dir):
+            os.mkdir(country_dir)
+        for region, new_rows in regions.items():
+            region_filename = os.path.join(country_dir, region + ".pickle")
+            try:
+                existing_rows: list[tuple] = pickle.load(open(region_filename, "rb"))
+                existing_rows.extend(new_rows)
+                print(
+                    "Write Results: {}/{} Len: {}".format(country, region, len(existing_rows))
+                )
+                pickle.dump(existing_rows, open(region_filename, "wb"))
+            except (OSError, IOError):
+                print(
+                    "Creating region .pickle file for {}/{} Len: {}...".format(
+                        country, region, len(new_rows)
+                    )
+                )
+                pickle.dump(new_rows, open(region_filename, "wb"))
```

Comment:
i dont see any synchronization here. if different processes are modifying or attempting to modify the same location at the same time bad things will occur. we have a number of ways around this though.

1. synchronization - we could do multiprocess synchronization. this will complicate the code and be slower
2. process sandboxing - each process can add another unique identify to the file names (you could use the processes PID). then they will never collide on write/update (though they may collide on create dir, you can handle this by checking if the dir already exists and swallowing the exception if it does). this will mean you'll need to work a bit harder in your aggregation script though,  you'll need to append multiple pickles together into your dataframe to make sure you have all the results from all processes

Reply:
- Added a UUID suffix for the filenames to prevent collisions
- Added the try-catch around mkdir for the country as you mentioned
- Changed aggregation to walk through a country's output dir first and combine all region files before turning it into a pandas DF / generating results